### ISSUE TITLE
Hold a RWMutex per Log, not per DB

### ISSUE THREAD

**Issue by badtuple** [badtuple/remits#19]
Right now, there's a single Mutex wrapped around the entire db:

```rust
// from main()
let db = Arc::new(Mutex::new(db::DB::new()));
```
This was mainly for simplicity while bootstrapping. Since adding a Message to a Log is the only place where mutation happens, we should instead have a RWLock on the Log itself. That would allow us to have:

- Multiple Logs adding messages at the same time
- Multiple reads happening from the same Log at the same time as long as a write isn't happening.
- Multiple connections not blocking each other trying to gain a db handle.

Once the persistence layer is in place, this will likely be our main non-IO bottleneck. It's not an issue yet, there's way more to do first...but I wanted to make sure to log the issue so we can come back to it.

If you are familiar with Rust and aren't scared of the borrow checker, this would be a great first issue with a really large impact. If unfamiliar with Rust, it's still a good first issue but realize you're going to get very familiar with the borrow checker.

**Comment by badtuple** [comment]
It occurs to me that since Logs are immutable, only the most recent page for each Log needs a write lock.  Even when writing to a Log, Iterators should be able to iterate over archived pages.

**Comment by badtuple** [comment]
Fixed by  #30
